### PR TITLE
java: fix java_home on macOS

### PR DESCRIPTION
**Pull request by jonchang**
https://github.com/Homebrew/brew/pull/10397 incorrectly changed what `java_home` pointed to on macOS. The path is correct on Linux, but on macOS, the Java home is actually in /usr/local/opt/openjdk/libexec/openjdk.jdk/Contents/Home to ensure compatibility with being linked to /Library/Java/JavaVirtualMachines per the openjdk formula caveat.

If a formula used e.g. `bin.write_jar_script` in its install block, this is what it looks like on macOS before and after this pull request.

```sh
#!/bin/bash
export JAVA_HOME="${JAVA_HOME:-/usr/local/opt/openjdk/libexec}"
exec "${JAVA_HOME}/bin/java"  -jar "/usr/local/Cellar/foo/1.0/libexec/foo.jar" "$@"
```

```sh
#!/bin/bash
export JAVA_HOME="${JAVA_HOME:-/usr/local/opt/openjdk/libexec/openjdk.jdk/Contents/Home}"
exec "${JAVA_HOME}/bin/java"  -jar "/usr/local/Cellar/foo/1.0/libexec/foo.jar" "$@"
```

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

### CODE REVIEW

**Review by carlocab — APPROVED** [review]
Thanks, @jonchang!

**Comment by carlocab** [comment]
```
Finished in 8 minutes 6 seconds (files took 5.65 seconds to load)
1022 examples, 6 failures

Failed examples:

rspec ./test/language/java_spec.rb:21 # Language::Java::java_home returns valid JAVA_HOME if version is specified
rspec ./test/language/java_spec.rb:26 # Language::Java::java_home returns valid JAVA_HOME if version is not specified
rspec ./test/language/java_spec.rb:45 # Language::Java::overridable_java_home_env returns java_home path if version specified
rspec ./test/language/java_spec.rb:50 # Language::Java::overridable_java_home_env returns java_home path if version is not specified
rspec ./test/language/java_spec.rb:38 # Language::Java::java_home_env returns java_home path if version is not specified
rspec ./test/language/java_spec.rb:33 # Language::Java::java_home_env returns java_home path if version specified
```

File: Library/Homebrew/test/language/java_spec.rb [hunk 1]
```diff
@@ -18,38 +18,38 @@
   end
 
   describe "::java_home" do
-    it "returns valid JAVA_HOME if version is specified" do
+    it "returns valid JAVA_HOME if version is specified", :needs_macos do
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/language/java_spec.rb:21]: This also requires having `openjdk` installed, right? Would be good to have a test for the Linux functionality too.

File: Library/Homebrew/extend/os/mac/language/java.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# typed: true
+# frozen_string_literal: true
+
+module Language
+  module Java
+    def self.java_home(version = nil)
+      find_openjdk_formula(version)&.opt_libexec&.join("openjdk.jdk/Contents/Home")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/mac/language/java.rb:7]:
```suggestion
      openjdk = find_openjdk_formula(version)
      return unless openjdk

      openjdk.opt_libexec/"openjdk.jdk/Contents/Home"
```

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for the fix!

File: Library/Homebrew/test/language/java_spec.rb (hunk 1 above)
Comment (carlocab) [reply Library/Homebrew/test/language/java_spec.rb:21]: Yes, I think it does. I was just duplicating the old test here. Unfortunately, it doesn't seem to work 😅

Comment (MikeMcQuaid) [reply Library/Homebrew/test/language/java_spec.rb:21]: Would need to add another stub to the `before do` to allow it to be an executable. That makes the test fairly pointless, though, if we're asserting on something we're stubbing to be the case. I'd suggest instead fixing up the equality value here and adding a `:needs_linux` equivalent test for the non-macOS functionality.

Comment (carlocab) [reply Library/Homebrew/test/language/java_spec.rb:21]: Hmmm, ok, gotcha. Will try that.

File: Library/Homebrew/test/language/java_spec.rb [hunk 3]
```diff
@@ -18,12 +18,22 @@
...
+      java_home = described_class.java_home("1.8+")
+      expect(java_home).to eql(f.opt_libexec.join("openjdk.jdk/Contents/Home").to_s)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/language/java_spec.rb:23]:
```suggestion
      expect(java_home).to eql(f.opt_libexec/"openjdk.jdk/Contents/Home").to_s)
```

File: Library/Homebrew/test/language/java_spec.rb [hunk 4]
```diff
@@ -18,12 +18,22 @@
...
+    end
+
+    it "returns valid JAVA_HOME if version is not specified", :needs_macos do
+      java_home = described_class.java_home
+      expect(java_home).to eql(f.opt_libexec.join("openjdk.jdk/Contents/Home").to_s)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/language/java_spec.rb:28]:
```suggestion
      expect(java_home).to eql(f.opt_libexec/"openjdk.jdk/Contents/Home").to_s)
```

File: Library/Homebrew/test/language/java_spec.rb [hunk 5]
```diff
@@ -32,24 +42,24 @@
   describe "::java_home_env" do
     it "returns java_home path if version specified" do
       java_home_env = described_class.java_home_env("1.8+")
-      expect(java_home_env[:JAVA_HOME]).to eql(f.opt_libexec.to_s)
+      expect(java_home_env[:JAVA_HOME]).to include(f.opt_libexec.to_s)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/language/java_spec.rb:45]: I think it'd be good to keep this as `eql` and add separate macOS/Linux tests if needed.

File: Library/Homebrew/test/language/java_spec.rb [hunk 6]
```diff
@@ -32,24 +42,24 @@
...
     end
 
     it "returns java_home path if version is not specified" do
       java_home_env = described_class.java_home_env
-      expect(java_home_env[:JAVA_HOME]).to eql(f.opt_libexec.to_s)
+      expect(java_home_env[:JAVA_HOME]).to include(f.opt_libexec.to_s)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/language/java_spec.rb:50]: I think it'd be good to keep this as `eql` and add separate macOS/Linux tests if needed.

File: Library/Homebrew/test/language/java_spec.rb [hunk 7]
```diff
@@ -32,24 +42,24 @@
...
     end
   end
 
   describe "::overridable_java_home_env" do
     it "returns java_home path if version specified" do
       overridable_java_home_env = described_class.overridable_java_home_env("1.8+")
-      expect(overridable_java_home_env[:JAVA_HOME]).to eql("${JAVA_HOME:-#{f.opt_libexec}}")
+      expect(overridable_java_home_env[:JAVA_HOME]).to include(f.opt_libexec.to_s)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/language/java_spec.rb:57]: I think it'd be good to keep this as `eql` and add separate macOS/Linux tests if needed.

File: Library/Homebrew/test/language/java_spec.rb [hunk 8]
```diff
@@ -32,24 +42,24 @@
...
     end
 
     it "returns java_home path if version is not specified" do
       overridable_java_home_env = described_class.overridable_java_home_env
-      expect(overridable_java_home_env[:JAVA_HOME]).to eql("${JAVA_HOME:-#{f.opt_libexec}}")
+      expect(overridable_java_home_env[:JAVA_HOME]).to include(f.opt_libexec.to_s)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/language/java_spec.rb:62]: I think it'd be good to keep this as `eql` and add separate macOS/Linux tests if needed.

**Review by MikeMcQuaid — COMMENTED** [review]
If it makes tests simpler I could see something like:

```ruby
let(:expected_home) do
  if OS.mac?
    f.opt_libexec.join("openjdk.jdk/Contents/Home").to_s
  else
    f.opt_libexec.to_s
  end
end
```

being a good fit; that would remove the need for all separate tests and the repetitions.

File: Library/Homebrew/test/language/java_spec.rb (hunk 5 above)
Comment (carlocab) [reply Library/Homebrew/test/language/java_spec.rb:45]: I agree. I'll follow that up in a separate PR since this is blocking some PRs in homebrew/core.

**Review by MikeMcQuaid — APPROVED** [review]
Fine with me to pull cleanup to a new PR 👍🏻

**Comment by jonchang** [comment]
Thanks for shepherding in this bug fix! I wasn't able to fix the tests last night so thanks for helping get this merged!

**Pull request merged**